Javascript Agent Documentation (#27)
undefinedlabs/scope-docs: docs/javascript-installation.md

},
...
```

### Cypress tests

If you want to instrument tests run by Cypress you need to add the scope-agent plugin.

An example [cypress.json](https://docs.cypress.io/guides/references/configuration.html#Folders-Files):

```
{
  ...
  "supportFile": "cypress/support/index.js",
  "pluginsFile": "cypress/plugins/index.js",
  ...
}
```

Where `cypress/support/index.js`:

```javascript
import { initializeCypress } from "@undefinedlabs/scope-agent";

initializeCypress();

... // any other configuration here
```

And `cypress/plugins/index.js`:

```javascript
const { initCypressPlugin } = require('@undefinedlabs/scope-agent')

module.exports = async (on, config) => {
  ... // any other plugin goes here
  const newConfig = await initCypressPlugin(on, config)
  return newConfig
}
```

## Environment variables

The following environment variables need to be configured when instrumenting your tests:

| Environment variable | Default value | Description                                            |
| -------------------- | ------------- | ------------------------------------------------------ |
| `$SCOPE_API_KEY`     |               | API key to use when sending data to Scope              |

> | `$SCOPE_HOST`        |               | API endpoint of the Scope installation to send data to |

Author: these are not following the standard but since it needs a new release of the agent, I'd rather change it later

Reviewer: Can we change it and re-release the agent?

Author: yeah I can do that

Author: done

---
id: javascript-installation
title: Scope Javascript Agent installation
sidebar_label: Installation
---

Installation of the Scope Agent is done via [npm](https://www.npmjs.com/package/@undefinedlabs/scope-agent).

> ```bash

Reviewer: We can use [docusaurus code tabs feature](https://docusaurus.io/docs/en/doc-markdown#language-specific-code-tabs) here

npm install @undefinedlabs/scope-agent
```

Or with `yarn`:

```bash
yarn add @undefinedlabs/scope-agent
```

## Usage

### Jest tests

If you want to instrument tests run by Jest, you need to create a `reporter`.

In `src/reporter.js`

```javascript
const Reporter = require("@undefinedlabs/scope-agent").JestReporter;

module.exports = Reporter;
```

Then you have to configure [jest reporter](https://jestjs.io/docs/en/configuration#reporters-array-modulename-modulename-options)
to point to this file.

For example:

In your `jest.config.js`:

```javascript
module.exports = {

>   ...

Reviewer: ```suggestion
  // ...
```

  reporters: ["<rootDir>/reporter.js"]

>   ...

Reviewer: ```suggestion
  // ...
```

};
```

#### Note for Create React App users

CRA does not allow `reporter` configuration for jest yet (more info in https://github.com/facebook/create-react-app/issues/2474) so you may not use the `jest` field in your `package.json`.

One alternative is running your test with inline configuration:

In your `package.json`:

> ```

Reviewer: ```suggestion
```json
```

I'd basically try to make the syntax be valid json syntax

...
"scripts": {
  ...
  "test:reporters": "react-scripts test --reporters=default --reporters=./src/reporter.js",
  ...
},
...
```

### Cypress tests

If you want to instrument tests run by Cypress you need to add the scope-agent plugin.

An example [cypress.json](https://docs.cypress.io/guides/references/configuration.html#Folders-Files):

```
{
  ...
  "supportFile": "cypress/support/index.js",
  "pluginsFile": "cypress/plugins/index.js",
  ...
}
```

Where `cypress/support/index.js`:

```javascript
import { initializeCypress } from "@undefinedlabs/scope-agent";

initializeCypress();

> ... // any other configuration here

Reviewer: ```suggestion
// ... any other configuration here
```

```

And `cypress/plugins/index.js`:

```javascript
const { initCypressPlugin } = require('@undefinedlabs/scope-agent')

module.exports = async (on, config) => {
  ... // any other plugin goes here
  const newConfig = await initCypressPlugin(on, config)
  return newConfig
}
```

> ## Environment variables

Reviewer: Do these work even in CRA? No need to prefix them as `REACT_APP_`?

Author: yeah they do. I believe `REACT_APP_` is just for environment variables consumed by the bundle, but we are consuming them in jest and cypress processes

---
id: javascript-installation
title: Scope Javascript Agent installation
sidebar_label: Installation
---

Installation of the Scope Agent is done via [npm](https://www.npmjs.com/package/@undefinedlabs/scope-agent).

<!--DOCUSAURUS_CODE_TABS-->
<!--npm-->

```bash
npm install --save-dev @undefinedlabs/scope-agent
```

<!--yarn-->

```bash
yarn add --dev @undefinedlabs/scope-agent
```

<!--END_DOCUSAURUS_CODE_TABS-->

## Instrumenting your tests

### Jest tests

If you want to instrument tests run by Jest, you need to create a `reporter`.

In `src/reporter.js`

```javascript
const Reporter = require("@undefinedlabs/scope-agent").JestReporter;

module.exports = Reporter;
```

Then you have to configure [jest reporter](https://jestjs.io/docs/en/configuration#reporters-array-modulename-modulename-options)
to point to this file.

For example:

In your `jest.config.js`:

```javascript
module.exports = {
  // ...

>   reporters: ["<rootDir>/reporter.js"]

Reviewer 2: You are using `src/reporter.js` as example path, but now you're using `<rootDir>/reporter.js`. It should be src/reporter.js again, right?

Author: you need to specify `<rootDir>` in `jest.config.js` https://jestjs.io/docs/en/configuration#rootdir-string. But there was indeed a mistake. It should be `<rootDir>/src/reporter.js` to keep consistency

  // ...
};
```

> #### Note for Create React App users

Reviewer 2: I'd say `Create React App` and give the command line first, a then the explanation. I don't care if CRA allows or not reporter config in the package.json

CRA does not allow `reporter` configuration for jest yet (more info in https://github.com/facebook/create-react-app/issues/2474) so you may not use the `jest` field in your `package.json`.

One alternative is running your test with inline configuration:

In your `package.json`:

```json
...
"scripts": {

>   "test:reporters": "react-scripts test --reporters=default --reporters=./src/reporter.js"

Reviewer 2: It should be more semantic, like: `test:scope` or `test:tracing` or even `test:debug`

}
...
```

### Cypress tests

If you want to instrument tests run by Cypress you need to add the scope-agent plugin.

> An example [cypress.json](https://docs.cypress.io/guides/references/configuration.html#Folders-Files):

Reviewer 2: ```suggestion
Add the following to [cypress.json](https://docs.cypress.io/guides/references/configuration.html#Folders-Files):
```

```json
{
  "supportFile": "cypress/support/index.js",
  "pluginsFile": "cypress/plugins/index.js"
}
```

Where `cypress/support/index.js`:

```javascript
import { initializeCypress } from "@undefinedlabs/scope-agent";

initializeCypress();

// ... any other configuration here
```

And `cypress/plugins/index.js`:

```javascript
const { initCypressPlugin } = require("@undefinedlabs/scope-agent");

module.exports = async (on, config) => {
  // ... any other plugin goes here
  const newConfig = await initCypressPlugin(on, config);
  return newConfig;
};
```

> ## Environment variables

Reviewer 2: I think we should put this section as `## Prerequisites` before anything else.

Author: while I would agree, this is the standard for the rest of the agents so I prefer to keep it this way
